New functions for CoolPropLib

friederikeboehm opened the pull request:
Props1SImulti new function in CoolProp and Props1SImulti and PropsSImulti to CoolPropLib

### Description of the Change

Function Props1SImulti was added to CoolProp. It is the Props1SI function but with multiple outputs possible, comparable to PropsSImulti.
Props1SImulti and PropsSImulti were added to CoolPropLib

### Benefits

Enables the user to acces multiple non-thermodynamic-state-dependend fluid properties at once.

### Possible Drawbacks

None

### Verification Process

Functions were tested using the Mathematica wrapper, see PR #2085

ibell commented:
What is the relationship between this PR and #2135 ? Should this be merged or closed?

friederikeboehm commented:
Both should be merged. I wanted to separate high-level and low-level functions (although I just realised #2135 has a high-level function as well...). Can join both PRs into one, if preferred.

ibell commented:
My preference is to separate the low-level and high-level changes (PR should be quite granular, and do one thing only, within reason). I will merge this one.

ibell commented:
Glad I thought about this, please refactor to move implementation into ``src/CoolProp.cpp``

friederikeboehm commented:
> Glad I thought about this, please refactor to move implementation into `src/CoolProp.cpp`

I'm sorry, I don't understand, what should I refactor? Props1SImulti is implemented in src/CoolProp.cpp

ibell commented:
The function should be implemented in ``src/CoolProp.cpp``, with STL types ``std::vector``, etc., and the function in ``CoolPropLib.cpp`` should *only* do C->C++ datatype conversions and pass to the function in ``src/CoolProp.cpp``, like all the other functions that are in ``src/CoolPropLib.cpp``

friederikeboehm commented on src/CoolProp.cpp:
```diff
@@ -834,6 +834,15 @@ double Props1SI(std::string FluidName, std::string Output) {
     }
 }
+
+std::vector<std::vector<double>> Props1SImulti(const std::vector<std::string>& Outputs, const std::string& backend, const std::vector<std::string>& fluids, const std::vector<double>& fractions) {
```
Implementation in CoolProp.cpp

friederikeboehm commented on src/CoolProp.cpp:
```diff
@@ -834,6 +834,15 @@ double Props1SI(std::string FluidName, std::string Output) {
+    std::vector<std::vector<double>> val1 = PropsSImulti(Outputs, "", zero_vector, "", zero_vector, backend, fluids, fractions);
+    // error handling is done in PropsSImulti, val1 will be an empty vector if an error occured
+    return val1;
+}
```
Implementation in CoolProp.cpp

friederikeboehm commented on src/CoolProp.cpp:
```diff
@@ -834,6 +834,15 @@ double Props1SI(std::string FluidName, std::string Output) {
         return val1;
     }
 }
+
+std::vector<std::vector<double>> Props1SImulti(const std::vector<std::string>& Outputs, const std::string& backend, const std::vector<std::string>& fluids, const std::vector<double>& fractions) {
+    std::vector<double> zero_vector(1, 0.);
+    std::vector<std::vector<double>> val1 = PropsSImulti(Outputs, "", zero_vector, "", zero_vector, backend, fluids, fractions);
+    // error handling is done in PropsSImulti, val1 will be an empty vector if an error occured
+    return val1;
+}
+
+
```
Implementation in CoolProp.cpp

friederikeboehm commented on src/CoolPropLib.cpp:
```diff
@@ -206,10 +206,97 @@ EXPORT_CODE double CONVENTION Props1SI(const char* FluidName, const char* Output
+            throw CoolProp::ValueError(
+              format("Length of fractions vector  [%d] is not equal to length of fluidNames vector [%d]", _fluidNames.size(), length_fractions));
+        std::vector<double> _fractions(fractions, fractions + length_fractions);
+        std::vector<std::vector<double>> _result = CoolProp::Props1SImulti(_outputs, backend, _fluidNames, _fractions);
```
Call to function in CoolProp.cpp

friederikeboehm commented:
> The function should be implemented in `src/CoolProp.cpp`, with STL types `std::vector`, etc., and the function in `CoolPropLib.cpp` should _only_ do C->C++ datatype conversions and pass to the function in `src/CoolProp.cpp`, like all the other functions that are in `src/CoolPropLib.cpp`

I thought that's exactly what I did? In CoolProp.cpp I implemented Props1SI and in CoolPropLib.cpp I call that function (and do a lot of error handling). I added comments at the according code changes in CoolProp.cpp and CoolPropLib.cpp. Please tell me, what I need to change as I don't understand what's wrong with my code.

ibell commented:
Sorry, I didn't read closely enough before
